pkg/nsx/services/ipaddressallocation/ipaddressallocation.go

```diff
@@ -0,0 +1,197 @@
+package ipaddressallocation
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var (
+	log                             = logger.Log
+	MarkedForDelete                 = true
+	ResourceTypeIPAddressAllocation = common.ResourceTypeIPAddressAllocation
+)
+
+type IPAddressAllocationService struct {
+	common.Service
+	ipAddressAllocationStore *IPAddressAllocationStore
+	VPCService               common.VPCServiceProvider
+}
+
+func InitializeIPAddressAllocation(service common.Service, vpcService common.VPCServiceProvider) (*IPAddressAllocationService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	ipAddressAllocationService := &IPAddressAllocationService{Service: service, VPCService: vpcService}
+	ipAddressAllocationService.ipAddressAllocationStore = &IPAddressAllocationStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPAddressAllocationCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBindingType(),
+	}}
+
+	tags := []model.Tag{
+		{Scope: String(common.TagScopeIPAddressAllocationCRUID)},
+	}
+	go ipAddressAllocationService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPAddressAllocation, tags, ipAddressAllocationService.ipAddressAllocationStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return ipAddressAllocationService, err
+	}
+	return ipAddressAllocationService, nil
+}
+
+func (service *IPAddressAllocationService) CreateOrUpdateIPAddressAllocation(obj *v1alpha1.IPAddressAllocation) (bool, error) {
+	nsxIPAddressAllocation, err := service.BuildIPAddressAllocation(obj)
+	if err != nil {
+		return false, err
+	}
+	existingIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
+		return false, err
+	}
+	log.V(1).Info("existing ipaddressallocation", "ipaddressallocation", existingIPAddressAllocation)
+	_ = common.CompareResource(IpAddressAllocationToComparable(existingIPAddressAllocation),
+		IpAddressAllocationToComparable(nsxIPAddressAllocation))
```

Comment:
It seems that we compared them but do nothing?

Reply:
oh, for test, forget to change it back